### PR TITLE
Officer Permissions

### PR DESCRIPTION
**Pull request by AJDurant**
Allows officer permissions to be added and revoked. Also provides macro for API urls.

### CODE REVIEW

**Comment by AJDurant** [comment]
There is potential for re-working editOfficer into a full edit form for all aspects? It could then also be used to create new officer positions

File: src/Classes/MyRadio/CoreUtils.php [hunk 1]
```diff
@@ -553,6 +553,22 @@ public static function getAllPermissions()
         );
     }
 
+    private static function comparePermission($perm1, $perm2)
```
Comment (LordAro) [inline src/Classes/MyRadio/CoreUtils.php]: doxymentation?

File: src/Classes/MyRadio/CoreUtils.php [hunk 2]
```diff
@@ -553,6 +553,22 @@ public static function getAllPermissions()
...
+        if ($perm1['value'] === $perm2['value']) {
+            return 0;
+        } elseif ($perm1['value'] < $perm2['value']) {
+            return -1;
+        } else {
+            return 1;
+        }
+    }
+
+    public static function diffPermissions($perms)
```
Comment (LordAro) [inline src/Classes/MyRadio/CoreUtils.php]: maybe "getMissingPermissions" instead? diffPermissions doesn't really tell you what the function does, imo

File: src/Templates/base.twig [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {% spaceless %}
-{% import 'macros.twig' as myury %}
+{% import 'macros.twig' as MyRadio %}
```
Comment (LordAro) [inline src/Templates/base.twig]: Perhaps should be lowercase?

Comment (AJDurant) [inline src/Templates/base.twig]: It's like a class name, tho if you want to change it you can

**Comment by LordAro** [comment]
this done now?

**Comment by AJDurant** [comment]
I've decided to actually make a full editOfficer page so i will do that as part of this PRQ first.

**Comment by AJDurant** [comment]
TODO: separate editOfficer into viewOfficer and new edit page

File: src/Classes/MyRadio/CoreUtils.php [hunk 4]
```diff
@@ -554,6 +554,34 @@ public static function getAllPermissions()
     }
 
     /**
+     * udiff function for permission value equality
+     * @param  array $perm1 permission value & description
+     * @param  array $perm2 permission value & description
+     * @return int          comparison result
+     */
+    private static function comparePermission($perm1, $perm2)
```
Comment (Flump5281) [inline src/Classes/MyRadio/CoreUtils.php]: Thinking about #319, should these potentially be moved out in that one? In which case if this is merged should that PR be updated to reflect it?

File: src/Classes/MyRadio/CoreUtils.php [hunk 5]
```diff
@@ -858,6 +886,18 @@ public static function getSafeHTML($dirty_html)
     }
 
     /**
+     * Returns lookup values for Status for a select box
+     * @return array
+     */
+    public static function getStatusLookup()
```
Comment (Flump5281) [inline src/Classes/MyRadio/CoreUtils.php]: Potentially Cache this?
We also have a specific form field type for college iirc - is it worth having one for this? I'm tempted to lean toward *not* having one for college and doing it like that.

File: src/Classes/ServiceAPI/MyRadio_Officer.php [hunk 6]
```diff
@@ -127,6 +144,25 @@ public static function getAllOfficerPositions()
         );
     }
 
+    /**
+     * Assigns and officership to the given member
```
Comment (Flump5281) [inline src/Classes/ServiceAPI/MyRadio_Officer.php]: an* :P

File: src/Classes/MyRadio/CoreUtils.php (hunk 4 above)
Comment (AJDurant) [inline src/Classes/MyRadio/CoreUtils.php]: Yeah, until that is merged stuff should still go in coreutils and separated out in that PR

File: src/Classes/ServiceAPI/MyRadio_Officer.php [hunk 7]
```diff
@@ -157,6 +193,23 @@ public function getName()
     }
 
     /**
+     * Sets the Name of this Officer Position
+     * @param String $name the new name of the officer
+     * @return MyRadio_Officer the updated officer object
+     */
+    public function setName($name)
```
Comment (Flump5281) [inline src/Classes/ServiceAPI/MyRadio_Officer.php]: You need to updateCacheObject when you do sets

File: src/Controllers/Profile/assignOfficer.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Assign a Member an Officership
+ *
+ * @package MyRadio_Profile
+ */
+
+use \MyRadio\MyRadio\CoreUtils;
+use \MyRadio\ServiceAPI\MyRadio_Officer;
+
+if ($_SERVER['REQUEST_METHOD'] === 'POST') {
```
Comment (Flump5281) [inline src/Controllers/Profile/assignOfficer.php]: Should we have a check to see if they are paid?

**Comment by AJDurant** [comment]
TODO:
- [ ] Revoke permissions link (api permissions?)
- [ ] Stand Down officer - move into API
- [ ] Test officer assign
- [ ] Update user cache object on assign / stand down

File: src/Classes/ServiceAPI/MyRadio_Officer.php [hunk 9]
```diff
@@ -181,6 +281,26 @@ public function getOrdering()
...
+    public function setOrdering($ordering)
+    {
+        if ($ordering !== $this->ordering) {
+            self::$db->query(
+                'UPDATE public.officer
+                SET ordering = $1
+                WHERE officerid=$2',
+                [$ordering, $this->getID()]
+            );
+            $this->ordering = $ordering;
```
Comment (Flump5281) [inline src/Classes/ServiceAPI/MyRadio_Officer.php]: Verify type here?

File: src/Classes/ServiceAPI/MyRadio_Officer.php [hunk 10]
```diff
@@ -265,9 +445,247 @@ public function getCurrentHolders()
...
+     */
+    public function addPermission($permissionid)
+    {
+        self::$db->query(
+            'INSERT INTO public.auth_officer
+            (officerid, lookupid)
+            VALUES ($1, $2)',
+            [$this->getID(), $permissionid]
+        );
+        $this->updateCacheObject();
```
Comment (Flump5281) [inline src/Classes/ServiceAPI/MyRadio_Officer.php]: You don't add the new permission to the local object, so the cache will still be wrong...

File: src/Classes/ServiceAPI/MyRadio_Officer.php [hunk 11]
```diff
@@ -265,9 +445,247 @@ public function getCurrentHolders()
...
+     */
+    public function revokePermission($permissionid)
+    {
+        self::$db->query(
+            'DELETE from public.auth_officer
+            WHERE officerid = $1
+            AND lookupid = $2',
+            [$this->getID(), $permissionid]
+        );
+        $this->updateCacheObject();
```
Comment (Flump5281) [inline src/Classes/ServiceAPI/MyRadio_Officer.php]: As above

File: src/Controllers/Profile/assignOfficer.php [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    //Submitted
+    $data = MyRadio_Officer::getAssignForm()->readValues();
+
+    $officer = MyRadio_Officer::getInstance($data['id']);
+
+    if ($data['member']->isCurrentlyPaid()) {
```
Comment (Flump5281) [inline src/Controllers/Profile/assignOfficer.php]: Potentially allow 'members' to not be paid?

File: src/Classes/ServiceAPI/MyRadio_Officer.php (hunk 10 above)
Comment (AJDurant) [inline src/Classes/ServiceAPI/MyRadio_Officer.php]: my thinking currently is just to update the whole permissions array - as finding a value in an array in an array seems more complex.

**Pull request merged**